### ISSUE TITLE
class extends basis class with generic compile error

### ISSUE THREAD

**Issue by HerrCai0907** [AssemblyScript/assemblyscript#1971]
Hello, 
Is this compile error reasonable? I'm not sure whether assemblyscript supports the way of following inheritance.

ERROR TS2322: Type '(this: assembly/index/B<f64,~lib/string/String>, f64) => void' is not assignable to type '(this: assembly/index/A<i32>, i32) => void'.

     foo(a:T):void{

```
class A<T>{
    foo(a:T):void{
        trace(`class A ${a}`)
    }
}
class B<T,V> extends A<T>{
    foo(a:T):void{
        trace(`class B ${a}`)
    }
}
export function test_foo():void{
    let v = new B<i32,f64>()
    let v2 = new B<i32,string>()
    let v3 = new B<f64,string>() // if comment this line, compile OK
    const v_usize = changetype<usize>(v)
    changetype<A<i32>>(v_usize).foo(1)
}
```

**Comment by HerrCai0907** [comment]
replace ```const v_usize = changetype<usize>(v); changetype<A<i32>>(v_usize).foo(1); ``` with ```(<A<i32>>v).foo(1);```, it still compiles error.

**Comment by willemneal** [comment]
I've seen this error before too and it is not reasonable. The presence of a different implementation of the class shouldn't affect the type resolution of another.  [I also found that if I change to casting instead of using `changetype`](https://www.assemblyscript.org/editor.html#IyFvcHRpbWl6ZT1zcGVlZCZydW50aW1lPXN0dWIKY2xhc3MgQTxUPnsKICAgIGZvbyhhOlQpOnZvaWR7CiAgICAgICAgdHJhY2UoYGNsYXNzIEEgJHthfWApCiAgICB9Cn0KY2xhc3MgQjxULFY+IGV4dGVuZHMgQTxUPnsKICAgIGZvbyhhOlQpOnZvaWR7CiAgICAgICAgdHJhY2UoYGNsYXNzIEIgJHthfWApCiAgICB9Cn0KZXhwb3J0IGZ1bmN0aW9uIHRlc3RfZm9vKCk6dm9pZHsKICAgIGxldCB2ID0gbmV3IEI8aTMyLGY2ND4oKQogICAgbGV0IHYyID0gbmV3IEI8aTMyLHN0cmluZz4oKQogICAgbGV0IHYzID0gbmV3IEI8ZjY0LHN0cmluZz4oKSAvLyBpZiBjb21tZW50IHRoaXMgbGluZSwgY29tcGlsZSBPSwogICAgKDxBPGkzMj4+KHYpKS5mb28oMSkKfQojIWh0bWwKPHRleHRhcmVhIGlkPSJvdXRwdXQiIHN0eWxlPSJoZWlnaHQ6IDEwMCU7IHdpZHRoOiAxMDAlIiByZWFkb25seT48L3RleHRhcmVhPgo8c2NyaXB0Pgpsb2FkZXIuaW5zdGFudGlhdGUobW9kdWxlX3dhc20sIHsgLyogaW1wb3J0cyAqLyB9KQogIC50aGVuKCh7IGV4cG9ydHMgfSkgPT4gewogICAgY29uc3Qgb3V0cHV0ID0gZG9jdW1lbnQuZ2V0RWxlbWVudEJ5SWQoJ291dHB1dCcpCiAgICBmb3IgKGxldCBpID0gMDsgaSA8PSAxMDsgKytpKSB7CiAgICAgIG91dHB1dC52YWx1ZSArPSBgZmliKCR7aX0pID0gJHtleHBvcnRzLmZpYihpKX1cbmAKICAgIH0KICB9KQo8L3NjcmlwdD4=) the same thing happens:

```ts
class A<T>{
    foo(a:T):void{
        trace(`class A ${a}`)
    }
}
class B<T,V> extends A<T>{
    foo(a:T):void{
        trace(`class B ${a}`)
    }
}
export function test_foo():void{
    let v = new B<i32,f64>()
    let v2 = new B<i32,string>()
    let v3 = new B<f64,string>() // if comment this line, compile OK
    (<A<i32>>v).foo(1)
}
```

**Comment by willemneal** [comment]
Digging into the compiler it seems that concrete instances of `A<T>`, e.g. `A<i32>` share's the same mapping of instances of all instances. All instances are checked against it regardless thus causing this error.

**Comment by trusktr** [comment]
I think I ran into an issue like this when `extend`ing from `Map<T>`'s and passing different types from the `Subclass<T>` to the `Map<T>`: the type system would complain about the type of an unrelated class being wrong.

In my particular case, I solved it by wrapping/containing a `Map<t>` in a non-subclass, instead of `extend`ing from `Map<T>`. This works but is more work to implement all the proxy methods (otherwise they would simply be inherited).
